### ISSUE TITLE
Question about focal loss implementation.

### ISSUE THREAD

**Issue by Markusgami** [xingyizhou/CenterNet#241]
CenterNet is a very good detection algorithm. It acheives 10% more accuracy on our dataset than our previous algorithm. Thanks for the work!
But I have a question on the focal loss implementation:

~~~~~~~~~~~~~~~~
def _neg_loss(pred, gt):
    ''' Modified focal loss. Exactly the same as CornerNet.
        Runs faster and costs a little bit more memory
      Arguments:
        pred (batch x c x h x w)
        gt_regr (batch x c x h x w)
    '''
    pos_inds = gt.eq(1).float()
    neg_inds = gt.lt(1).float()

    neg_weights = torch.pow(1 - gt, 4)

    loss = 0

    pos_loss = torch.log(pred) * torch.pow(1 - pred, 2) * pos_inds
    neg_loss = torch.log(1 - pred) * torch.pow(pred, 2) * neg_weights * neg_inds

    num_pos = pos_inds.float().sum()
    pos_loss = pos_loss.sum()
    neg_loss = neg_loss.sum()

    if num_pos == 0:
        loss = loss - neg_loss
    else:
        loss = loss - (pos_loss + neg_loss) / num_pos

    return loss
~~~~~~~~~~~~~~

if num_pos == 0:
        loss = loss - neg_loss --->lossA
else:
        loss = loss - (pos_loss + neg_loss) / num_pos  --->lossB

If there is any positive in this batch of data, this loss function will return lossB even though maybe most of the data in batch has no positive..
So for a img, it will have different affect on loss accordding to different batch composition.
I don't understand if this will have any good or bad effect on training.

**Comment by Markusgami** [comment]
RegL1Loss has the same problem(maybe not a problem?)

**Comment by xingyizhou** [comment]
That's a good point. I am not aware of that. Will investigate it later.

**Comment by Markusgami** [comment]
I'm training the Network on "Correct version" of loss now. We will see if it's getting better or just nothing happend..
I noticed this implementation is from CornerNet..So their code is not same with their paper..

![image](https://user-images.githubusercontent.com/26424470/62347124-40639500-b52b-11e9-99d9-841dc2066cf4.png)

**Comment by shuoshuoa** [comment]
@Markusgami What is the effect of your verification?I have also noticed this problem.

**Comment by realwill** [comment]
@Markusgami is the 'correct version' of loss work?

**Comment by Markusgami** [comment]
Sorry.. Only thing I know now is that "corrected loss" gives comparable results...I trained with some other changes to the CenterNet and GPU is always busy, so I still don't know the isolated effect on the loss function

**Comment by DeeperDeeper** [comment]
@Markusgami 
could you share your "corrected loss" implementation ?

**Comment by Markusgami** [comment]
def _neg_loss(pred, gt):
  ''' Modified focal loss. Exactly the same as CornerNet.
      Runs faster and costs a little bit more memory
    Arguments:
      pred (batch x c x h x w)
      gt_regr (batch x c x h x w)
  '''
  pos_inds = gt.eq(1).float()
  neg_inds = gt.lt(1).float()

  neg_weights = torch.pow(1 - gt, 4)

  lossA = 0
  lossB = 0

  pos_loss = torch.log(pred) * torch.pow(1 - pred, 2) * pos_inds
  neg_loss = torch.log(1 - pred) * torch.pow(pred, 2) * neg_weights * neg_inds

  num_pos  = pos_inds.float().sum((1, 2, 3))
  pos_loss = pos_loss.sum((1, 2, 3))
  neg_loss = neg_loss.sum((1, 2, 3))


  mask = 1- (num_pos==0)

  lossA = lossA - neg_loss
  lossB = lossB - (pos_loss + neg_loss) / (num_pos+1e-4)

  loss = mask.float() * lossB + (1-mask).float() * lossA
  return loss


class RegL1Loss(nn.Module):
  def __init__(self):
    super(RegL1Loss, self).__init__()

  def forward(self, output, mask, ind, target):
    pred = _tranpose_and_gather_feat(output, ind)
    mask = mask.unsqueeze(2).expand_as(pred).float()
    loss2 = torch.abs((pred * mask - target * mask)).sum((1, 2))
    loss2 = loss2 / (mask.sum((1, 2)) + 1e-4)
    return loss2

CenterNet will take mean for the loss in base_trainer.py

**Comment by DeeperDeeper** [comment]
@Markusgami, so here if I understand correctly, you did not sum over the batch axis, so this function will return a batch of loss (a batch of scalar ??), is it right? and after that you take the mean of 3 losses:
`loss = mean(L_keypoints + L_offset + 0.1* L_size)?`

**Comment by Markusgami** [comment]
Yeah..Before the change, the loss function return one loss value for the whole batch directly.
After my change, the loss function will return a vector which contains one loss value for one image..You need to take the mean across the batch before do the backward.

**Comment by DeeperDeeper** [comment]
Thanks @Markusgami, did you get good results with these changes?
